File: problems-adapt/0200_count-grid-islands/solution_bfs.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {

    public int countGridIslands(String[][] grid) {
        int rows = grid.length;
        if (rows == 0) return 0;
        int cols = grid[0].length;
        boolean[][] visited = new boolean[rows][cols];
        int count = 0;
        int[][] dirs = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c].equals("1") && !visited[r][c]) {
                    count++;
                    Deque<int[]> queue = new ArrayDeque<>();
                    queue.offer(new int[] { r, c });
                    visited[r][c] = true;
                    while (!queue.isEmpty()) {
                        int[] cell = queue.poll();
                        for (int[] d : dirs) {
                            int nx = cell[0] + d[0],
                                ny = cell[1] + d[1];
                            if (
                                nx >= 0 &&
                                nx < rows &&
                                ny >= 0 &&
                                ny < cols &&
                                grid[nx][ny].equals("1") &&
                                !visited[nx][ny]
                            ) {
                                visited[nx][ny] = true;
                                queue.offer(new int[] { nx, ny });
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}

File: problems-adapt/0000-0100/0071_count-grid-islands/solution_bfs.py
from collections import deque


class Solution:
    def countGridIslands(self, grid: list[list[str]]) -> int:
        if not grid:
            return 0
        rows, cols = len(grid), len(grid[0])
        visited = [[False] * cols for _ in range(rows)]
        count = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and not visited[r][c]:
                    count += 1
                    queue = deque([(r, c)])
                    visited[r][c] = True
                    while queue:
                        x, y = queue.popleft()
                        for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                            nx, ny = x + dx, y + dy
                            if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == "1" and not visited[nx][ny]:
                                visited[nx][ny] = True
                                queue.append((nx, ny))
        return count
